run_galaxy_workflow.py

```diff
@@ -131,104 +119,308 @@ def get_input_data_id(file, wf):
+                    " '{}' parameter step label is not present in the workflow definition".format(step_label_p))
+            logging.error("{} parameter step label is not present in the workflow definition".format(step_label_p))
+    logging.info("Validation of labels: OK")
+
+
+def validate_input_labels(wf_json, inputs):
+    """
+    Check that all input datasets in the workflow have labels, and that those labels are available in the inputs yaml.
+    Raises an exception if those cases are not fulfilled.
+
+    :param wf_json:
+    :param inputs:
+    :return:
+    """
+
+    for step, step_content in wf_json['steps'].items():
+        if step_content['type'] == 'data_input':
+            if step_content['label'] is None:
+                raise ValueError("Input step {} in workflow has no label set.".format(str(step)))
+
+            if step_content['label'] not in inputs:
+                raise ValueError("Input step {} label {} is not present in the inputs YAML file provided."
+                                 .format(str(step), step_content['label']))
+
+
+def read_yaml_file(yaml_path):
+    """
+    Reads a YAML file safely.
+
+    :param yaml_path:
+    :return: dictionary object from YAML content.
+    """
+    stream = open(yaml_path, "r")
+    return yaml.safe_load(stream)
+
+
+def validate_file_exists(inputs):
+    """
+    Checks that paths exists in the local file system
+
+    :param inputs: dictionary with inputs
+    :return:
+    """
+    for input_key, input_content in inputs.items():
+        if 'path' in input_content and not os.path.isfile(input_content['path']):
+            raise ValueError("Input file {} does not exist for input label {}".format(input_content['path'], input_key))
+
+
+def produce_versions_file(gi, workflow_from_json, path):
+    """
+    Produces a tool versions file for the workflow run.
+
+    :param gi:
+    :param workflow_from_json:
+    :return:
+    """
+
+    tools_dict = []
+
+    f = open(file=path, mode="w")
+    f.write("\t".join(["Analysis", "Software", "Version", "Citation"])+"\n")
+
+    for key, step in sorted(workflow_from_json['steps'].items(), reverse=True):
+        # Input steps won't have tool ids, and we only need each tool once.
+        if step['tool_id'] is not None and step['tool_id'] not in tools_dict:
+            tool = gi.tools.show_tool(step['tool_id'])
+            label = step['label'] if step['label'] is not None else tool['name']
+            url = ""
+            if tool['tool_shed_repository'] is not None:
+                ts_meta = tool['tool_shed_repository']
+                url = "https://{}/view/{}/{}/{}".format(ts_meta['tool_shed'], ts_meta['owner'], ts_meta['name'], ts_meta['changeset_revision'])
+            f.write("\t".join([label, tool['name'], tool['version'], url])+"\n")
+            tools_dict.append(step['tool_id'])
+            # tools_dict[step['tool_id']] = {'name': tool['name'], 'version': tool['version']}
+    f.flush()
+    f.close()
```

Comment:
You can save a few lines and use the `with open(filename, 'w') as f:` structure that handles the file closing automatically (also when something goes wrong).